File: 9/9b.py
with open("9.txt") as file:
    lines = file.read().split('\n')

    visitedSpaces = {'0,0': True}

    class Coordinates:
        def __init__(self, x, y):
            self.x = x
            self.y = y

    knots = [Coordinates(0, 0) for _ in range(10)]

    for line in lines:
        (direction, count) = line.split(' ')
        for i in range(int(count)):
            head = knots[0]
            if direction == 'L':
                head.x -= 1
            elif direction == 'R':
                head.x += 1
            elif direction == 'U':
                head.y -= 1
            elif direction == 'D':
                head.y += 1

            for j in range(1, len(knots)):
                xDiff = knots[j - 1].x - knots[j].x
                yDiff = knots[j - 1].y - knots[j].y

                if abs(xDiff) == 2:
                    knots[j].x += int(xDiff/2)
                    if abs(yDiff) == 1:
                        knots[j].y += yDiff
                if abs(yDiff) == 2:
                    knots[j].y += int(yDiff/2)
                    if abs(xDiff) == 1:
                        knots[j].x += xDiff

            tail = knots[-1]
            visitedSpaces[f'{tail.y},{tail.x}'] = True

    print(len(visitedSpaces))

File: 9/9b.js
const fs = require('fs');

const input = fs.readFileSync('./9.txt', 'utf8');

const lines = input.split(/\r?\n/);

const visitedSpaces = { '0,0': true };

const knots = [];
for (let i = 0; i < 10; i++) {
  knots.push({ x: 0, y: 0 });
}

for (const line of lines) {
  const [dir, count] = line.split(' ');
  for (let i = 0; i < count; i++) {
    const head = knots[0];
    if (dir === 'L') head.x -= 1;
    else if (dir === 'R') head.x += 1;
    else if (dir === 'U') head.y -= 1;
    else if (dir === 'D') head.y += 1;

    for (let j = 1; j < knots.length; j++) {
      const xDiff = knots[j - 1].x - knots[j].x;
      const yDiff = knots[j - 1].y - knots[j].y;

      // knots can have diff=abs(2) in both directions, except with head
      // pair of diffs that cause moves can be:
      // abs(2) in one direction and abs(0) or abs(1) in the other;
      // or abs(2) in both directions
      if (Math.abs(xDiff) === 2) {
        knots[j].x += xDiff / 2;
        if (Math.abs(yDiff) === 1) {
          knots[j].y += yDiff;
        }
      }
      if (Math.abs(yDiff) === 2) {
        knots[j].y += yDiff / 2;
        if (Math.abs(xDiff) === 1) {
          knots[j].x += xDiff;
        }
      }
    }

    const tail = knots[knots.length - 1];
    visitedSpaces[tail.y + ',' + tail.x] = true;
  }
}

console.log(Object.keys(visitedSpaces).length);
